### PR TITLE
Implement dtx.recover frame

### PR DESCRIPTION
**Pull request by a5anka**
The dtx.recover frame is used to query the broker node on distributed transaction in prepared state.

### CODE REVIEW

File: modules/andes-core/client/src/main/java/org/wso2/andes/client/XAResource_0_9_1.java [hunk 1]
```diff
@@ -241,7 +241,20 @@ public int prepare(Xid xid) throws XAException {
...
+        List<Xid> xidList;
+        try {
+            xidList = session.recoverDtxTransactions();
+        } catch (FailoverException | AMQException e) {
+            XAException xaException = new XAException("Error while recovering dtx sessions.");
+            xaException.initCause(e);
+            throw xaException;
+        }
+
+        return xidList.toArray(new Xid[xidList.size()]);
```
Comment (sdkottegoda) [inline modules/andes-core/client/src/main/java/org/wso2/andes/client/XAResource_0_9_1.java]: Can we return the list from inside the try?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/server/handler/DtxRecoverHandler.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import org.wso2.andes.server.AMQChannel;
+import org.wso2.andes.server.protocol.AMQProtocolSession;
+import org.wso2.andes.server.state.AMQStateManager;
+import org.wso2.andes.server.state.StateAwareMethodListener;
+import org.wso2.andes.server.txn.DtxNotSelectedException;
+
+import java.util.ArrayList;
+import javax.transaction.xa.Xid;
+
+public class DtxRecoverHandler implements StateAwareMethodListener<DtxRecoverBodyImpl> {
```
Comment (ramith) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/server/handler/DtxRecoverHandler.java]: a class comment?

File: modules/andes-core/client/src/main/java/org/wso2/andes/client/XAResource_0_9_1.java [hunk 3]
```diff
@@ -241,7 +241,19 @@ public int prepare(Xid xid) throws XAException {
...
+            LOGGER.debug("recover dtx branches with prepared state");
+        }
+
+        try {
+            List<Xid> xidList = session.recoverDtxTransactions();
+            return xidList.toArray(new Xid[xidList.size()]);
+
+        } catch (FailoverException | AMQException e) {
+            XAException xaException = new XAException("Error while recovering dtx sessions.");
+            xaException.initCause(e);
```
Comment (lankavitharana) [inline modules/andes-core/client/src/main/java/org/wso2/andes/client/XAResource_0_9_1.java]: Is this some standard or can't we add the exception in Exception constructor?(without calling init method?)

File: modules/andes-core/client/src/main/java/org/wso2/andes/client/handler/ClientMethodDispatcherImpl_0_91.java [hunk 4]
```diff
@@ -249,6 +251,15 @@ public boolean dispatchDtxPrepareOk(DtxPrepareOkBody body, int channelId) throws
     }
 
     @Override
+    public boolean dispatchDtxRecoverOk(DtxRecoverOkBody body, int channelId) throws AMQException {
+        if (LOGGER.isDebugEnabled())
+        {
```
Comment (lankavitharana) [inline modules/andes-core/client/src/main/java/org/wso2/andes/client/handler/ClientMethodDispatcherImpl_0_91.java]: can we move this to the line above? like if (LOGGER.isDebugEnabled()) {

File: modules/andes-core/common/src/main/java/org/wso2/andes/dtx/XidImpl.java [hunk 5]
```diff
@@ -24,13 +24,14 @@
...
+import java.io.Serializable;
 import java.util.Arrays;
 import javax.transaction.xa.Xid;
 
 /**
  * Implements javax.transaction.dtx.Xid
  */
-public class XidImpl implements Xid
+public class XidImpl implements Xid, Serializable
 {
```
Comment (lankavitharana) [inline modules/andes-core/common/src/main/java/org/wso2/andes/dtx/XidImpl.java]: can we move this to the line above as well?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/server/handler/DtxRecoverHandler.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+import org.wso2.andes.server.txn.DtxNotSelectedException;
+
+import java.util.ArrayList;
+import javax.transaction.xa.Xid;
+
+/**
+ * Handler class for dtx.recover
+ */
+public class DtxRecoverHandler implements StateAwareMethodListener<DtxRecoverBodyImpl> {
+    private static DtxRecoverHandler _instance = new DtxRecoverHandler();
```
Comment (lankavitharana) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/server/handler/DtxRecoverHandler.java]: can we remove "_" from the start of the variable name?

File: modules/andes-core/client/src/main/java/org/wso2/andes/client/XAResource_0_9_1.java (hunk 3 above)
Comment (a5anka) [inline modules/andes-core/client/src/main/java/org/wso2/andes/client/XAResource_0_9_1.java]: XAException does not have a constructor to pass both error message and the exception. Therefore we had to do this.

File: modules/andes-core/client/src/main/java/org/wso2/andes/client/handler/ClientMethodDispatcherImpl_0_91.java (hunk 4 above)
Comment (a5anka) [inline modules/andes-core/client/src/main/java/org/wso2/andes/client/handler/ClientMethodDispatcherImpl_0_91.java]: :+1:

File: modules/andes-core/common/src/main/java/org/wso2/andes/dtx/XidImpl.java (hunk 5 above)
Comment (a5anka) [inline modules/andes-core/common/src/main/java/org/wso2/andes/dtx/XidImpl.java]: :+1:

File: modules/andes-core/broker/src/main/java/org/wso2/andes/server/handler/DtxRecoverHandler.java (hunk 6 above)
Comment (a5anka) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/server/handler/DtxRecoverHandler.java]: :+1:

File: modules/andes-core/client/src/main/java/org/wso2/andes/client/handler/ClientMethodDispatcherImpl_0_91.java (hunk 4 above)
Comment (a5anka) [inline modules/andes-core/client/src/main/java/org/wso2/andes/client/handler/ClientMethodDispatcherImpl_0_91.java]: Fixed in https://github.com/wso2/andes/pull/794

File: modules/andes-core/common/src/main/java/org/wso2/andes/dtx/XidImpl.java (hunk 5 above)
Comment (a5anka) [inline modules/andes-core/common/src/main/java/org/wso2/andes/dtx/XidImpl.java]: Fixed in https://github.com/wso2/andes/pull/794

File: modules/andes-core/broker/src/main/java/org/wso2/andes/server/handler/DtxRecoverHandler.java (hunk 6 above)
Comment (a5anka) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/server/handler/DtxRecoverHandler.java]: Fixed in https://github.com/wso2/andes/pull/794

**Pull request merged**